Content.Server/Cloning/CloningSystem.Utility.cs

```diff
@@ -0,0 +1,357 @@
+        clonePod.ActivelyCloning = false;
+    }
+
+    /// <summary>
+    ///     And now we turn it over to Chef Pod to make soup!
+    /// </summary>
+    private void EndFailedCloning(EntityUid uid, CloningPodComponent clonePod)
+    {
+        if (clonePod.BodyContainer.ContainedEntity is { Valid: true } entity)
+        {
+            if (TryComp<PhysicsComponent>(entity, out var physics)
+                && TryComp<BloodstreamComponent>(entity, out var bloodstream))
+                MakeAHugeMess(uid, physics, bloodstream);
+            else MakeAHugeMess(uid);
+
+            QueueDel(entity);
+        }
+        else MakeAHugeMess(uid);
+
+        clonePod.FailedClone = false;
+        clonePod.CloningProgress = 0f;
+        UpdateStatus(uid, CloningPodStatus.Idle, clonePod);
+        if (HasComp<EmaggedComponent>(uid))
+        {
+            _audio.PlayPvs(clonePod.ScreamSound, uid);
+            Spawn(clonePod.MobSpawnId, Transform(uid).Coordinates);
+        }
+
+        if (!HasComp<EmaggedComponent>(uid))
+            _material.SpawnMultipleFromMaterial(_random.Next(1, (int) (clonePod.UsedBiomass / 2.5)), clonePod.RequiredMaterial, Transform(uid).Coordinates);
+
+        clonePod.UsedBiomass = 0;
+        clonePod.ActivelyCloning = false;
+    }
+
+    /// <summary>
+    ///     The body coming out of the machine isn't guaranteed to even be a Humanoid.
+    ///     This function makes sure the body is "Human Playable", with no funny business.
+    /// </summary>
+    private void CleanupCloneComponents(EntityUid uid, EntityUid bodyToClone, bool forceOldProfile, bool doMetempsychosis)
+    {
+        if (forceOldProfile
+            && TryComp<PsionicComponent>(bodyToClone, out var psionic))
+        {
+            var newPsionic = _serialization.CreateCopy(psionic, null, false, true);
+            AddComp(uid, newPsionic, true);
+        }
+
+        if (TryComp<LanguageKnowledgeComponent>(bodyToClone, out var oldKnowLangs))
+        {
+            var newKnowLangs = _serialization.CreateCopy(oldKnowLangs, null, false, true);
+            AddComp(uid, newKnowLangs, true);
+        }
+
+
+        if (TryComp<LanguageSpeakerComponent>(bodyToClone, out var oldSpeakLangs))
+        {
+            var newSpeakLangs = _serialization.CreateCopy(oldSpeakLangs, null, false, true);
+            AddComp(uid, newSpeakLangs, true);
+        }
+
+        if (doMetempsychosis)
+            EnsureComp<PsionicComponent>(uid);
+
+        EnsureComp<SpeechComponent>(uid);
+        EnsureComp<DamageForceSayComponent>(uid);
+        EnsureComp<EmotingComponent>(uid);
+        EnsureComp<MindContainerComponent>(uid);
+        EnsureComp<SSDIndicatorComponent>(uid);
+        RemComp<ReplacementAccentComponent>(uid);
+        RemComp<MonkeyAccentComponent>(uid);
+        RemComp<SentienceTargetComponent>(uid);
+        RemComp<GhostTakeoverAvailableComponent>(uid);
+        _tag.AddTag(uid, "DoorBumpOpener");
```

Comment:
It gets bigger every time I look at it... There needs to be a better solution.

Reply:
The only other alternative I can think of is to use the method that BuildAMimicWorkshop calls, but that method has a significant weakness in that it CANNOT be used with an Abstract or NoSpawn entity, so whatever "Reference Entity" is used to workshop all the "These are components a player character requires", must itself be a totally valid entity to spawn. Even then it won't work for Removing components, so we'd be bringing in an awful solution to replace 5 EnsureComps at most.

Reply:
We could... Technically add a trio of ComponentRegistries to the ClonePodComponent? CopyComponents, EnsureComponents, RemoveComponents?